src/utils/samlUtils.ts

```diff
@@ -0,0 +1,1176 @@
+                      )
+                    )
+                  )
+                )
+              )
+            ),
+            E.chain(() =>
+              pipe(
+                E.fromOption(() => new Error("Missing AuthnStatement"))(
+                  O.fromNullable(
+                    Assertion.getElementsByTagNameNS(
+                      SAML_NAMESPACE.ASSERTION,
+                      "AuthnStatement"
+                    ).item(0)
+                  )
+                ),
+                E.chain(
+                  E.fromPredicate(
+                    PR.not(isEmptyNode),
+                    () => new Error("Empty AuthnStatement")
+                  )
+                ),
+                E.chain(AuthnStatement =>
+                  pipe(
+                    E.fromOption(() => new Error("Missing AuthnContext"))(
+                      O.fromNullable(
+                        AuthnStatement.getElementsByTagNameNS(
+                          SAML_NAMESPACE.ASSERTION,
+                          "AuthnContext"
+                        ).item(0)
+                      )
+                    ),
+                    E.chain(
+                      E.fromPredicate(
+                        PR.not(isEmptyNode),
+                        () => new Error("Empty AuthnContext")
+                      )
+                    ),
+                    E.chain(AuthnContext =>
+                      pipe(
+                        E.fromOption(
+                          () => new Error("Missing AuthnContextClassRef")
+                        )(
+                          O.fromNullable(
+                            AuthnContext.getElementsByTagNameNS(
+                              SAML_NAMESPACE.ASSERTION,
+                              "AuthnContextClassRef"
+                            ).item(0)
+                          )
+                        ),
+                        E.chain(
+                          E.fromPredicate(
+                            PR.not(isEmptyNode),
+                            () => new Error("Empty AuthnContextClassRef")
+                          )
+                        ),
+                        E.chain(
+                          E.fromPredicate(
+                            AuthnContextClassRef =>
+                              AuthnContextClassRef.textContent?.trim() ===
+                                SPID_LEVELS.SpidL1 ||
+                              AuthnContextClassRef.textContent?.trim() ===
+                                SPID_LEVELS.SpidL2 ||
+                              AuthnContextClassRef.textContent?.trim() ===
+                                SPID_LEVELS.SpidL3,
+                            () =>
+                              new Error("Invalid AuthnContextClassRef value")
+                          )
+                        ),
+                        E.map(AuthnContextClassRef =>
+                          AuthnContextClassRef.textContent?.trim()
+                        ),
```

Comment:
We can optimize this code doing the map before the chain
```suggestion
                        E.map(AuthnContextClassRef =>
                          AuthnContextClassRef.textContent?.trim()
                        ),
                        E.chain(
                          E.fromPredicate(
                            AuthnContextClassRef =>
                              AuthnContextClassRef === SPID_LEVELS.SpidL1 ||
                              AuthnContextClassRef === SPID_LEVELS.SpidL2 ||
                              AuthnContextClassRef === SPID_LEVELS.SpidL3,
                            () =>
                              new Error("Invalid AuthnContextClassRef value")
                          )
                        ),
```

Reply:
nice catch! fixed!